File: src/exe/serve/web/route/games/owned.rs
//! Owned game release routes.

use std::collections::HashMap;

use axum::extract::{Query, State};
use axum::http::StatusCode;
use diesel::prelude::*;
use diesel_async::RunQueryDsl;
use media::game::Game;
use media::game::owned::{Body, Owned, Patch, Row};
use utoipa_axum::router::OpenApiRouter as Router;
use utoipa_axum::routes;
use uuid::Uuid;

use super::super::query::Order;
use crate::axum::extract::{Error, Json, Path};
use crate::db::{self, Pool, Uuid as DbUuid};
use crate::schema::{games as g, games_owned as t};

pub fn router() -> Router<Pool> {
    Router::new()
        .routes(routes!(list, create))
        .routes(routes!(fetch, update, modify, remove))
}

/// Load the game with the given ID.
async fn load_game(conn: &mut db::Conn, id: Uuid) -> Result<Game, Error> {
    g::table
        .filter(g::id.eq(DbUuid::from(id)))
        .select(g::all_columns)
        .first(conn)
        .await
        .inspect_err(|err| tracing::error!("{err}"))
        .map_err(Error::from)
}

/// Load games for a set of IDs, keyed by ID.
async fn load_games_for(conn: &mut db::Conn, ids: &[DbUuid]) -> Result<HashMap<Uuid, Game>, Error> {
    let rows: Vec<Game> = g::table
        .filter(g::id.eq_any(ids))
        .select(g::all_columns)
        .load(conn)
        .await
        .inspect_err(|err| tracing::error!("{err}"))
        .map_err(Error::from)?;
    Ok(rows.into_iter().map(|game| (game.id, game)).collect())
}

/// Sort field for owned game releases.
#[derive(Clone, Copy, Debug, Default)]
#[derive(serde::Deserialize)]
#[derive(utoipa::ToSchema)]
#[serde(rename_all = "lowercase")]
enum Sort {
    /// Sort by platform.
    #[default]
    System,
    /// Sort by model.
    Model,
}

/// Query parameters for listing owned game releases.
#[derive(Clone, Debug, Default)]
#[derive(serde::Deserialize)]
#[derive(utoipa::IntoParams)]
struct Params {
    /// Filter by game ID.
    game: Option<Uuid>,
    /// Filter by platform.
    system: Option<String>,
    /// Field to sort by.
    #[param(inline)]
    sort: Option<Sort>,
    /// Sort direction.
    #[param(inline)]
    order: Option<Order>,
    /// Maximum number of results.
    limit: Option<i64>,
    /// Number of results to skip.
    offset: Option<i64>,
}

/// List owned game releases.
#[utoipa::path(
    get,
    path = "/",
    tag = "games/owned",
    params(Params),
    responses((status = 200, body = Vec<Owned>)),
)]
async fn list(
    State(db): State<Pool>,
    Query(params): Query<Params>,
) -> Result<Json<Vec<Owned>>, Error> {
    let mut conn = db::get_conn(&db).await.map_err(Error::from)?;

    let mut query = t::table.select(t::all_columns).into_boxed();

    if let Some(game) = params.game {
        query = query.filter(t::game.eq(DbUuid::from(game)));
    }
    if let Some(system) = params.system {
        query = query.filter(t::system.eq(system));
    }

    let rows: Vec<Row> = {
        let q = match (
            params.sort.unwrap_or_default(),
            params.order.unwrap_or_default(),
        ) {
            (Sort::System, Order::Asc) => query.order_by(t::system.asc()),
            (Sort::System, Order::Desc) => query.order_by(t::system.desc()),
            (Sort::Model, Order::Asc) => query.order_by(t::model.asc()),
            (Sort::Model, Order::Desc) => query.order_by(t::model.desc()),
        };
        if let Some(limit) = params.limit {
            q.limit(limit)
                .offset(params.offset.unwrap_or(0))
                .load(&mut conn)
                .await
        } else {
            q.load(&mut conn).await
        }
    }
    .inspect_err(|err| tracing::error!("{err}"))
    .map_err(Error::from)?;

    // Resolve games
    //
    // NOTE: Games are cloned rather than removed from the map, since
    // several releases may reference the same game.
    let ids: Vec<DbUuid> = rows.iter().map(|row| DbUuid::from(row.game)).collect();
    let games = load_games_for(&mut conn, &ids).await?;
    let rows = rows
        .into_iter()
        .filter_map(|row| {
            let game = games.get(&row.game).cloned()?;
            Some(row.resolve(game))
        })
        .collect();

    Ok(Json(rows))
}

/// Fetch an owned game release by ID.
#[utoipa::path(
    get,
    path = "/{id}",
    tag = "games/owned",
    params(("id" = Uuid, Path)),
    responses((status = 200, body = Owned), (status = 404)),
)]
async fn fetch(State(db): State<Pool>, Path(id): Path<Uuid>) -> Result<Json<Owned>, Error> {
    let mut conn = db::get_conn(&db).await.map_err(Error::from)?;
    let row = t::table
        .select(t::all_columns)
        .filter(t::id.eq(DbUuid::from(id)))
        .first::<Row>(&mut conn)
        .await
        .optional()
        .map_err(Error::from)?
        .ok_or(Error::NotFound)?;
    let game = load_game(&mut conn, row.game).await?;
    Ok(Json(row.resolve(game)))
}

/// Create an owned game release.
#[utoipa::path(
    post,
    path = "/",
    tag = "games/owned",
    security(("BearerAuth" = [])),
    request_body(content = inline(Body)),
    responses((status = 201, body = Uuid), (status = 500)),
)]
async fn create(
    State(db): State<Pool>,
    Json(body): Json<Body>,
) -> Result<(StatusCode, Json<Uuid>), Error> {
    let mut conn = db::get_conn(&db).await.map_err(Error::from)?;
    let id = Uuid::new_v4();
    diesel::insert_into(t::table)
        .values((
            t::id.eq(DbUuid::from(id)),
            t::game.eq(DbUuid::from(body.game)),
            t::system.eq(&body.system),
            t::region.eq(&body.region),
            t::model.eq(&body.model),
            t::revision.eq(&body.revision),
            t::serial.eq(&body.serial),
            t::complete.eq(body.complete.unwrap_or(false)),
            t::modified.eq(body.modified.unwrap_or(false)),
        ))
        .execute(&mut conn)
        .await
        .inspect_err(|err| tracing::error!("{err}"))
        .map_err(Error::from)?;
    Ok((StatusCode::CREATED, Json(id)))
}

/// Update an owned game release.
#[utoipa::path(
    put,
    path = "/{id}",
    tag = "games/owned",
    params(("id" = Uuid, Path)),
    security(("BearerAuth" = [])),
    request_body(content = inline(Body)),
    responses((status = 200, body = Owned), (status = 404)),
)]
async fn update(
    State(db): State<Pool>,
    Path(id): Path<Uuid>,
    Json(body): Json<Body>,
) -> Result<Json<Owned>, Error> {
    let mut conn = db::get_conn(&db).await.map_err(Error::from)?;
    let uid = DbUuid::from(id);
    let n = diesel::update(t::table.filter(t::id.eq(uid)))
        .set((
            t::game.eq(DbUuid::from(body.game)),
            t::system.eq(&body.system),
            t::region.eq(&body.region),
            t::model.eq(&body.model),
            t::revision.eq(&body.revision),
            t::serial.eq(&body.serial),
            t::complete.eq(body.complete.unwrap_or(false)),
            t::modified.eq(body.modified.unwrap_or(false)),
        ))
        .execute(&mut conn)
        .await
        .inspect_err(|err| tracing::error!("{err}"))
        .map_err(Error::from)?;
    if n == 0 {
        return Err(Error::NotFound);
    }
    let row = t::table
        .select(t::all_columns)
        .filter(t::id.eq(uid))
        .first::<Row>(&mut conn)
        .await
        .map_err(Error::from)?;
    let game = load_game(&mut conn, row.game).await?;
    Ok(Json(row.resolve(game)))
}

/// Modify an owned game release.
#[utoipa::path(
    patch,
    path = "/{id}",
    tag = "games/owned",
    params(("id" = Uuid, Path)),
    security(("BearerAuth" = [])),
    request_body(content = inline(Patch)),
    responses((status = 200, body = Owned), (status = 404)),
)]
async fn modify(
    State(db): State<Pool>,
    Path(id): Path<Uuid>,
    Json(body): Json<Patch>,
) -> Result<Json<Owned>, Error> {
    let mut conn = db::get_conn(&db).await.map_err(Error::from)?;
    let uid = DbUuid::from(id);
    // Apply present fields
    if !body.is_empty() {
        let n = diesel::update(t::table.filter(t::id.eq(uid)))
            .set((
                body.game.map(|v| t::game.eq(DbUuid::from(v))),
                body.system.map(|v| t::system.eq(v)),
                body.region.map(|v| t::region.eq(v)),
                body.model.map(|v| t::model.eq(v)),
                body.revision.map(|v| t::revision.eq(v)),
                body.serial.map(|v| t::serial.eq(v)),
                body.complete.map(|v| t::complete.eq(v)),
                body.modified.map(|v| t::modified.eq(v)),
            ))
            .execute(&mut conn)
            .await
            .inspect_err(|err| tracing::error!("{err}"))
            .map_err(Error::from)?;
        if n == 0 {
            return Err(Error::NotFound);
        }
    }
    let row = t::table
        .select(t::all_columns)
        .filter(t::id.eq(uid))
        .first::<Row>(&mut conn)
        .await
        .map_err(Error::from)?;
    let game = load_game(&mut conn, row.game).await?;
    Ok(Json(row.resolve(game)))
}

/// Delete an owned game release.
#[utoipa::path(
    delete,
    path = "/{id}",
    tag = "games/owned",
    params(("id" = Uuid, Path)),
    security(("BearerAuth" = [])),
    responses((status = 204), (status = 404)),
)]
async fn remove(State(db): State<Pool>, Path(id): Path<Uuid>) -> Result<StatusCode, Error> {
    let mut conn = db::get_conn(&db).await.map_err(Error::from)?;
    let n = diesel::delete(t::table.filter(t::id.eq(DbUuid::from(id))))
        .execute(&mut conn)
        .await
        .inspect_err(|err| tracing::error!("{err}"))
        .map_err(Error::from)?;
    if n > 0 {
        Ok(StatusCode::NO_CONTENT)
    } else {
        Err(Error::NotFound)
    }
}

File: src/web/routes/games/owned.ts
//! Owned game routes.
//!
//! Releases, consoles, and accessories are one table discriminated by kind:
//! they record the same hardware detail and the same ordered list of games. The
//! kind paths are sugar over the `kind` query parameter, while an item is
//! addressed by its identifier alone.

import { and, asc, desc, eq, inArray, like } from "drizzle-orm";
import type { SQLiteColumn } from "drizzle-orm/sqlite-core";
import { Elysia, t } from "elysia";

import {
    bool,
    BY_KIND,
    choice,
    define,
    ident,
    list,
    nullable,
    page,
    uuid,
    type Fields,
} from "../../../models";
import * as db from "../../../sql";
import * as schema from "../../../sql/schema";
import {
    bare,
    created,
    empty,
    fail,
    failed,
    Id,
    json,
    NO_CONTENT,
    operation,
} from "../../reply";

/** What an owned item is. */
export const KINDS = ["release", "console", "extra"] as const;

export type Kind = (typeof KINDS)[number];

/** How a request body may name the games an item carries. */
export type Refs = { game?: string | undefined; games?: string[] | undefined };

/**
 * Resolves the games an item carries.
 *
 * A list is taken as given, while a lone identifier stands for a list of one.
 * Naming neither leaves the references alone.
 */
export function refs(given: Refs) {
    if (given.games !== undefined) return given.games;
    if (given.game !== undefined) return [given.game];
    return undefined;
}

/** Columns holding collection state, which default to false. */
const FLAGS = ["complete", "modified"];

/** Referenced game, as embedded in a resolved item. */
const Game = BY_KIND.get("game")!.item;

/** Column declarations, in wire order. */
const FIELDS: Fields = {
    id: uuid("Unique identifier."),
    kind: t.Unsafe<never>({
        description: "Item kind.",
        enum: [...KINDS],
    }),
    title: nullable(t.String(), "Title."),
    platform: nullable(t.String(), "Platform."),
    region: nullable(t.String(), "Region code."),
    model: nullable(t.String(), "Model name."),
    revision: nullable(t.String(), "Hardware revision."),
    serial: nullable(t.String(), "Serial number."),
    variant: nullable(t.String(), "Variant."),
    complete: bool("Complete-in-box status."),
    modified: bool("Hardware modification status."),
};

/** Columns a request body may set. The kind comes from the path. */
const COLUMNS = Object.keys(FIELDS).filter(
    (name) => name !== "id" && name !== "kind",
);

const { kind: discriminant, ...columns } = FIELDS;

/** Stored item, holding the columns of one owned thing. */
const Item = define(
    "OwnedGame",
    t.Object(columns, { description: "Owned game item." }),
);

/** Owned item, with its game references resolved. */
const Owned = t.Object(
    {
        kind: discriminant!,
        item: Item,
        games: list(Game, "Included games."),
    },
    { description: "Owned game record." },
);

const body: Fields = {
    title: t.Optional(FIELDS.title!),
    game: t.Optional(uuid("Included game; a single-element `games`.")),
    games: t.Optional(list(uuid("Game reference."), "Included games.")),
};
for (const name of COLUMNS) {
    if (name === "title") continue;
    body[name] = FLAGS.includes(name)
        ? t.Optional(nullable(t.Boolean(), FIELDS[name]!.description ?? ""))
        : t.Optional(FIELDS[name]!);
}

const Body = t.Object(body, { description: "Request body." });
const Patch = t.Partial(t.Object(body), {
    description: "Partial request body.",
});

const TAG = "games/owned";

const SORT = ["title", "platform", "model"];

export function router(cxn: db.Cxn) {
    const params = t.Object({ id: ident });

    /** Answers a request for an item that is not on file. */
    const missing = () => fail("not_found", "No owned game item with that ID.");
    const filters = {
        q: t.Optional(
            t.String({
                description: "Search title (case-insensitive substring).",
            }),
        ),
        game: t.Optional(
            t.String({
                format: "uuid",
                description: "Filter by included game ID.",
            }),
        ),
        platform: t.Optional(t.String({ description: "Filter by platform." })),
        sort: choice(SORT, "Field to sort by."),
        ...page,
    };
    const query = t.Object({
        kind: choice([...KINDS], "Filter by item kind."),
        ...filters,
    });

    const ref = schema.games_owned_ref;

    /** Loads the games referenced by each item, keyed by item ID. */
    const games = async (ids: string[]) => {
        const rows = await cxn
            .select({ owned: ref.owned, game: schema.games })
            .from(ref)
            .innerJoin(schema.games, eq(schema.games.id, ref.game))
            .where(inArray(ref.owned, ids))
            .orderBy(ref.owned, ref.idx);

        const out = new Map<string, unknown[]>();
        for (const row of rows) {
            const held = out.get(row.owned);
            if (held) held.push(row.game);
            else out.set(row.owned, [row.game]);
        }
        return out;
    };

    /** Replaces the games referenced by an item, preserving list order. */
    type Tx = Parameters<Parameters<db.Cxn["transaction"]>[0]>[0];
    const relate = (tx: Tx, owned: string, held: string[]) => {
        tx.delete(ref).where(eq(ref.owned, owned)).run();
        held.forEach((game, idx) => {
            tx.insert(ref)
                .values({ owned, game, idx })
                .onConflictDoNothing()
                .run();
        });
    };

    /** Wraps each row in the record envelope, resolving its game references. */
    const resolve = async (rows: { id: string }[]) => {
        const held = await games(rows.map((row) => row.id));
        return rows.map((row) => {
            const { kind, ...item } = row as Record<string, unknown>;
            return { kind, item, games: held.get(row.id) ?? [] };
        });
    };

    const fetch = async (id: string) => {
        const rows = await cxn
            .select()
            .from(schema.games_owned)
            .where(eq(schema.games_owned.id, id));
        return rows.length
            ? (await resolve(rows as { id: string }[]))[0]
            : undefined;
    };

    /** Splits a request body into column values and game references. */
    const split = (given: Record<string, unknown>, fill: boolean) => {
        const values: Record<string, unknown> = {};
        for (const name of COLUMNS) {
            const blank = FLAGS.includes(name) ? false : null;
            if (name in given) values[name] = given[name] ?? blank;
            else if (fill) values[name] = blank;
        }
        return { values, held: refs(given as Refs) };
    };

    type Query = {
        q?: string | undefined;
        game?: string | undefined;
        platform?: string | undefined;
        sort?: string | undefined;
        order?: string | undefined;
        limit?: number | undefined;
        offset?: number | undefined;
    };

    /** Lists items, optionally constrained to one kind. */
    const search = async (only: Kind | undefined, args: Query) => {
        // Apply filters
        const where = [
            only ? eq(schema.games_owned.kind, only) : undefined,
            args.q ? like(schema.games_owned.title, `%${args.q}%`) : undefined,
            args.game
                ? inArray(
                      schema.games_owned.id,
                      cxn
                          .select({ owned: ref.owned })
                          .from(ref)
                          .where(eq(ref.game, args.game)),
                  )
                : undefined,
            args.platform
                ? eq(schema.games_owned.platform, args.platform)
                : undefined,
        ];

        // Sort and paginate
        const columns: Record<string, SQLiteColumn> = {
            title: schema.games_owned.title,
            platform: schema.games_owned.platform,
            model: schema.games_owned.model,
        };
        const column = columns[args.sort ?? "title"]!;
        const sorted = cxn
            .select()
            .from(schema.games_owned)
            .where(and(...where))
            .orderBy((args.order === "asc" ? asc : desc)(column));
        const rows = await (args.limit === undefined
            ? sorted
            : sorted.limit(args.limit).offset(args.offset ?? 0));

        return resolve(rows as { id: string }[]);
    };

    /** Records a new item of the given kind. */
    const insert = (kind: Kind, given: unknown) => {
        const id = crypto.randomUUID();
        const { values, held } = split(given as Record<string, unknown>, true);
        cxn.transaction((tx) => {
            tx.insert(schema.games_owned)
                .values({
                    id,
                    kind,
                    ...values,
                } as typeof schema.games_owned.$inferInsert)
                .run();
            relate(tx, id, held ?? []);
        });
        return created(id);
    };

    const found = json(Owned);
    const listed = json(t.Array(Owned));

    return new Elysia({ prefix: "/games/owned", name: "owned:games" })
        .get(
            "",
            ({ query: args }) => search(args.kind as Kind | undefined, args),
            {
                query,
                detail: operation({
                    tag: TAG,
                    id: "listGamesOwned",
                    about: "List owned game items.",
                    responses: { 200: listed },
                }),
            },
        )
        .get(
            "/:id",
            async ({ params: args }) =>
                (await fetch(args.id)) ?? missing(),
            {
                params,
                detail: operation({
                    tag: TAG,
                    id: "fetchGamesOwned",
                    about: "Fetch an owned game item by ID.",
                    responses: { 200: found, 404: failed },
                }),
            },
        )
        .put(
            "/:id",
            async ({ params: args, body: given }) => {
                const { values, held } = split(
                    given as Record<string, unknown>,
                    true,
                );
                const changed = cxn.transaction((tx) => {
                    const res = tx
                        .update(schema.games_owned)
                        .set(values)
                        .where(eq(schema.games_owned.id, args.id))
                        .run();
                    if (db.affected(res)) relate(tx, args.id, held ?? []);
                    return db.affected(res);
                });
                if (!changed) return missing();
                return (await fetch(args.id))!;
            },
            {
                params,
                body: Body,
                parse: "json",
                detail: operation({
                    tag: TAG,
                    id: "updateGamesOwned",
                    about: "Update an owned game item.",
                    write: true,
                    responses: { 200: found, 404: failed },
                }),
            },
        )
        .patch(
            "/:id",
            async ({ params: args, body: given }) => {
                const { values, held } = split(
                    given as Record<string, unknown>,
                    false,
                );

                // Apply present fields
                if (Object.keys(values).length || held) {
                    cxn.transaction((tx) => {
                        if (Object.keys(values).length) {
                            tx.update(schema.games_owned)
                                .set(values)
                                .where(eq(schema.games_owned.id, args.id))
                                .run();
                        }
                        if (held) relate(tx, args.id, held);
                    });
                }

                return (await fetch(args.id)) ?? missing();
            },
            {
                params,
                body: Patch,
                parse: "json",
                detail: operation({
                    tag: TAG,
                    id: "modifyGamesOwned",
                    about: "Modify an owned game item.",
                    write: true,
                    responses: { 200: found, 404: failed },
                }),
            },
        )
        .delete(
            "/:id",
            async ({ params: args }) => {
                const res = await cxn
                    .delete(schema.games_owned)
                    .where(eq(schema.games_owned.id, args.id));
                return db.affected(res) ? empty(NO_CONTENT) : missing();
            },
            {
                params,
                detail: operation({
                    tag: TAG,
                    id: "removeGamesOwned",
                    about: "Delete an owned game item.",
                    write: true,
                    responses: { 204: bare, 404: failed },
                }),
            },
        )
        .get(`/releases`, ({ query: args }) => search("release", args), {
            query: t.Object(filters),
            detail: operation({
                tag: TAG,
                id: "listReleases",
                about: "List owned releases.",
                responses: { 200: listed },
            }),
        })
        .post(`/releases`, ({ body: given }) => insert("release", given), {
            body: Body,
            parse: "json",
            detail: operation({
                tag: TAG,
                id: "createRelease",
                about: "Create a release.",
                write: true,
                responses: { 201: json(Id), 500: failed },
            }),
        })
        .get(`/consoles`, ({ query: args }) => search("console", args), {
            query: t.Object(filters),
            detail: operation({
                tag: TAG,
                id: "listConsoles",
                about: "List owned consoles.",
                responses: { 200: listed },
            }),
        })
        .post(`/consoles`, ({ body: given }) => insert("console", given), {
            body: Body,
            parse: "json",
            detail: operation({
                tag: TAG,
                id: "createConsole",
                about: "Create a console.",
                write: true,
                responses: { 201: json(Id), 500: failed },
            }),
        })
        .get(`/extras`, ({ query: args }) => search("extra", args), {
            query: t.Object(filters),
            detail: operation({
                tag: TAG,
                id: "listExtras",
                about: "List owned extras.",
                responses: { 200: listed },
            }),
        })
        .post(`/extras`, ({ body: given }) => insert("extra", given), {
            body: Body,
            parse: "json",
            detail: operation({
                tag: TAG,
                id: "createExtra",
                about: "Create an extra.",
                write: true,
                responses: { 201: json(Id), 500: failed },
            }),
        });
}
